File: typescript/fenwick.ts
/*A Fenwick tree or binary indexed tree (BIT) is a data structure that can
efficiently update values and calculate prefix sums in an array of values.*/

class Fenwick {
    private nums: number[];
    
    constructor(n) {
        this.nums = Array(n+1).fill(0); 
    }
    
    add(k, delta) {
        for (++k; k < this.nums.length; k += k & -k)
            this.nums[k] += delta; 
    }
    
    query(k) {
        let ans = 0; 
        for (++k; k; k -= k & -k) 
            ans += this.nums[k]; 
        return ans; 
    }
}

File: java/fenwick.java
class Fenwick {
    /*Fenwick tree for prefix sum query*/
    private int[] nums; 

    public Fenwick(int n) {
        nums = new int[n+1]; 
    }

    /*Update tree upon adding delta to kth element.*/
    public void add(int k, int delta) {
        for (++k; k < nums.length; k += k & -k)
            nums[k] += delta; 
    }

    /*Return the prefix sum up to kth index (inclusive).*/
    public long query(int k) {
        long ans = 0; 
        for (++k; k > 0; k -= k & -k) 
            ans += nums[k]; 
        return ans; 
    }
}
